### ISSUE TITLE
REPL segfaults when typing ß

### ISSUE THREAD

**Issue by konsumlamm** [racket/racket#3223]
The REPL segfaults for me, when i type `ß` 10 times (the message is in german, in case you're wondering).

```
$ racket
Welcome to Racket v6.11.
> ÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃÃ^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@malloc(): memory corruption
Abgebrochen (Speicherabzug geschrieben)
```

My OS is `Linux Mint 19.3 Tricia`.

EDIT: Installing `readline-gpl` fixed the issue. However, this probably shouldn't segfault nonetheless.

**Comment by jbclements** [comment]
If you have a few minutes, would you consider checking to see if this bug is present in a more recent version of racket?

Also, based on the discussion in #1956 (https://github.com/racket/racket/issues/1956), it sounds like this might actually be a bug in libedit?

**Comment by samth** [comment]
I can confirm that this happens for me with libedit, and not with xrepl turned off or with readline-gpl.

**Comment by samth** [comment]
Based on reading some of the libedit man page, it seems like the following stack trace:

```
Thread 1 "racket3m" received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) where
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff7c5f859 in __GI_abort () at abort.c:79
#2  0x00007ffff7cca3ee in __libc_message (action=action@entry=do_abort, fmt=fmt@entry=0x7ffff7df4285 "%s\n")
    at ../sysdeps/posix/libc_fatal.c:155
#3  0x00007ffff7cd247c in malloc_printerr (str=str@entry=0x7ffff7df6670 "double free or corruption (out)") at malloc.c:5347
#4  0x00007ffff7cd4120 in _int_free (av=0x7ffff7e25b80 <main_arena>, p=0x555555a66770, have_lock=<optimized out>) at malloc.c:4314
#5  0x00007fffeea8bc61 in ?? () from /usr/lib/x86_64-linux-gnu/libedit.so.2
#6  0x00007fffeea7a680 in ?? () from /usr/lib/x86_64-linux-gnu/libedit.so.2
#7  0x00007fffeea79d84 in el_wgets () from /usr/lib/x86_64-linux-gnu/libedit.so.2
#8  0x00007fffeea75267 in el_gets () from /usr/lib/x86_64-linux-gnu/libedit.so.2
#9  0x00007fffeea8803e in readline () from /usr/lib/x86_64-linux-gnu/libedit.so.2
#10 0x00007ffff7e5581e in ffi_call_unix64 () from /usr/lib/x86_64-linux-gnu/libffi.so.6
#11 0x00007ffff7e551ef in ffi_call () from /usr/lib/x86_64-linux-gnu/libffi.so.6
#12 0x00005555557d7d7e in finish_ffi_call (cif=<optimized out>, c_func=<optimized out>, cfoff=<optimized out>, nargs=<optimized out>, 
    ivals=<optimized out>, avalues=<optimized out>, offsets=0x7fffffffb270, p=0x7fffffffb190)
    at ../../../racket/gc2/../src/../../foreign/foreign.c:3599
#13 0x00005555557e09a1 in ffi_do_call (argv=<optimized out>, self=<optimized out>, argc=<optimized out>)
    at ../../../racket/gc2/../src/../../foreign/foreign.c:3706
#14 0x00005555557e0e59 in ffi_do_call_after_stack_check (argc=<optimized out>, argv=<optimized out>, self=<optimized out>)
    at ../../../racket/gc2/../src/../../foreign/foreign.c:3765
```

indicates that the libedit wide-character functions are being called -- `el_wgets` is a wide-character version.

**Comment by samth** [comment]
Furthermore, I have a version of libedit from 2019 and apparently wide characters were made the only option in 2016. 

I wonder if we need to update our treatment of libedit based on that change?

**Comment by konsumlamm** [comment]
> If you have a few minutes, would you consider checking to see if this bug is present in a more recent version of racket?

It is still present in Racket v7.7 for me. Installing `readline-gpl` again fixes the issue.

**Comment by konsumlamm** [comment]
Should this be closed, since it's fixed by installing `readline-gpl`, or stay open, because of the segfault?

**Comment by samth** [comment]
I think this is a real bug.

**Comment by samth** [comment]
tldr: the problem is our use of `rl_getc_function`.

Consider the following program, which is a Racket version of the [editline example](https://github.com/troglobit/editline#example). 
```
(require readline/rktrl)
(let l ()
  (define v (readline "cli>"))
  (unless (eof-object? v)
    (writeln v)
    (l)))
```

If you paste ß into that repl, you get the same problem. However, if you bind `readline` yourself, the problem goes away (and unicode input such as ß looks nice again, yay!). The difference is the use of `rl_getc_function` -- comment that out from `rktrl.rkt` and the bugs go away. Furthermore, the content of that function is very suspicious, since it treats bytes as the right thing to return from something like `getc`, which is unlikely to be right in world with Unicode and multibyte encodings.

Unfortunately, I haven't yet found a good workaround for this; see https://lists.gnu.org/archive/html/bug-readline/2015-02/msg00026.html for some more discussion.

**Comment by samth** [comment]
One thing to note for anyone reading this -- the `editline` library is primarily used here is at https://www.thrysoee.dk/editline/ and not the one I linked to above.

**Comment by samth** [comment]
Also, here's Ruby discovering why we use `rl_getc_function`: https://bugs.ruby-lang.org/issues/5539 and here's them discovering the problem we have: https://bugs.ruby-lang.org/issues/9204

**Comment by rmculpepper** [comment]
If I make the following change at the end of readline-lib/readline/rktrl.rkt, then Unicode input works on recent libedit for me:

    (set-ffi-obj! "rl_getc_function" libreadline (_fun #:keep keep-alive! _pointer -> _int)
                  (lambda (_)
                    (define next-char (read-char real-input-port))
                    (if (eof-object? next-char) -1 (char->integer next-char))))

That is, read and return a whole character at a time.

That agrees with `_getc_function` in libedit/src/readline.c, which calls `rl_getc_function` and then casts the result to a `wchar_t`.

The code above might not work with old versions of libedit, and I don't know (yet?) what the right test for compatibility would be. And this change doesn't work for readline.

**Comment by samth** [comment]
A flag for whether we want wide characters seems like the right thing then, and setting it to #f for readline, or detecting readline dynamically, would make this work.

**Comment by rmculpepper** [comment]
Yes, but what I meant was that I don't know how to calculate that flag for libedit.

The libedit changelog (http://thrysoee.dk/editline/) says that the wchar_t behavior was introduced as an option in version "0:35:0" (2010-04-24) and became mandatory in version "0:54:0" (2016-06-18). But those numbers don't seem to correspond with the value of `rl_readline_version`, which is currently `#x402` (?!).

But I guess we can just change it for all version of editline. Broken for old versions is better than broken for new versions.

**Comment by samth** [comment]
There's already some old version detection, for some history feature, see the `old-editline` flag.

**Comment by samth** [comment]
Actually, a better test would be for the wide character functions from the non-readline API.

**Comment by rmculpepper** [comment]
Neither of those really detect the right feature, although `el_wgets` seems like a good way to distinguish libedit from libreadline.

**Comment by samth** [comment]
Both of those characters now work for me. What do you get from `racketcs -W debug@ffi-lib`?

**Comment by rmculpepper** [comment]
Both of those work for me.

I tested this in gnome-terminal with Encoding set to "Unicode - UTF-8". I started racket from a shell where $LANG is "en_US.UTF-8". My system's libedit is /usr/lib/x86_64-linux-gnu/libedit.so.2.0.63. Are your settings different?

**Comment by samth** [comment]
My settings are basically identical, except there's `tmux` and `mosh` between racket & bash on one side and Gnome Terminal on the other.

**Comment by sorawee** [comment]
```
Welcome to Racket v7.8.0.5.
ffi-lib: loaded "libcurses.dylib"
ffi-lib: loaded "libtermcap.dylib"
ffi-lib: failed for (ffi-lib "libedit" '("2.11")), tried:
  #<path:/Users/sorawee/Library/Racket/development/lib/libedit.2.11.dylib> (no such file)
  #<path:/Users/sorawee/Library/Racket/development/lib/libedit> (no such file)
  #<path:/Users/sorawee/git/racket/racket/lib/libedit.2.11.dylib> (no such file)
  #<path:/Users/sorawee/git/racket/racket/lib/libedit> (no such file)
  "libedit.2.11.dylib" (using OS library search path)
  "libedit" (using OS library search path)
  #<path:/Users/sorawee/git/racket-repos/libedit.2.11.dylib> (no such file)
  #<path:/Users/sorawee/git/racket-repos/libedit> (no such file)
ffi-lib: loaded "libedit.3.dylib"
```

I'm on Mac btw. Could this be the difference?

**Comment by samth** [comment]
Yes, that could definitely be the difference. But the thing to figure out is what "libedit.3" corresponds to (and whether there are any other libedit dylibs that we should use instead).

**Comment by sorawee** [comment]
https://pewpewthespells.com/blog/osx_readline.html looks relevant. Sigh.

**Comment by samth** [comment]
A few other things to note for posterity here: Debian Jessie (released 2015) shipped a version prior to wide characters being unconditionally enabled. Even that version had `#x402` as the `rl_readline_version` (so no point in looking there). For that very old version, wide character support caused the `hist_convert` function to be available, so we could detect that (but it's probably not worth doing more now).

**Comment by sorawee** [comment]
Interestingly, after installing `readline-gpl`, unicode still doesn't work, even though the loaded `dylib` is changed:

```
Welcome to Racket v7.8.0.5.
ffi-lib: loaded "libcurses.dylib"
ffi-lib: loaded "libtermcap.dylib"
ffi-lib: loaded "libreadline.dylib"
```

**Comment by samth** [comment]
Reading https://opensource.apple.com/source/libedit/ suggests that it was last updated in 2012. Maybe `readline-gpl` should come with a binary on for macOS.

**Comment by samth** [comment]
On the other hand, https://opensource.apple.com/source/libedit/libedit-55/ has a config.h that suggests wide chars were enabled then. Is the conditional that @rmculpepper added true for you? If you change it to be always true, do things work?

**Comment by sorawee** [comment]
Nope, it doesn't help, which makes sense: as I mentioned previously, the fix by @rmculpepper (already) has an effect. Typing ß no longer crashes the REPL.

**Comment by samth** [comment]
How does Python do with similar characters?

**Comment by sorawee** [comment]
It works fine... Perhaps they didn't use `rl_getc_function`?

Note that `python3` reports that it's using GNU readline, though there's a possibility that it got confused because MacOS symlinked `libreadline.dylib` to `libedit.3.dylib`. `python2` though reports that it's using `libedit`.

**Comment by samth** [comment]
Do things work better for you if `rl_getc_function` isn't used at all?

**Comment by sorawee** [comment]
Correct. If I comment 

```
(set-ffi-obj! "rl_getc_function" ...)
```

out, then unicode works fine.

**Comment by samth** [comment]
The use of `rl_getc_function` was added based on this discussion: https://github.com/racket/gnats-bugs/blob/master/all/13350

Maybe one of the other options discussed there would work better, or the previous approach involving `rl_event_hook`.

**Comment by rmculpepper** [comment]
As far as I can tell, based on a look through the code Sam linked to, setting `rl_getc_function` effectively disables Unicode handling. The problem is in readline.c in the function `_getc_function`, which calls `rl_getc_function` and casts its `int` result to a `char`. In contrast, the default `read_char` (defined in read.c) can return a full `wchar_t` (`Char` is a typedef for `wchar_t`).

I installed libedit via homebrew and ran racket with `PLT_READLINE_LIB` set to its path. With that version, Unicode input works for me. So it might be better to make a nonbroken libedit available for Mac OS (either bundled with Racket or in a package) rather than trying to work around the system version.

**Comment by mflatt** [comment]
I agree with @rmculpepper's suggestion to add a newer libedit build to the Racket native-library packages for Mac OS, and I've done that.
